Activate can be called several times

gthvn1 opened the pull request:
As it is idempotent we just need to return if the VDI is already activate.

ydirson committed:
zfs-ng: implement sr-create 1/4: SR.create

SR can then be disposed of using "xe sr-forget" and "zpool destroy".

Underlying zpool can be either separately created and given to
SR.create, or a set of devices can be specified to create a new zpool
on.

Notes:
- `SR.attach` will be necessary to complete `sr-create`: while from a
  calling user PoV things appear to have gone fine, SMlog reveals a
  call to `SR.attach`, which naturally failed as not yet implemented.
  Any attempt to `xe pbd-plug` naturally also fails.  See
  xapi-project/xen-api#5532.
- there is no plugin feature to shield from SR.* calls

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: base infrastructure

Bare minimum for a plugin, no operation implemented.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: implement Volume.destroy

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
WIP zfs-ng: implement Volume.resize

FIXME: it seems valid to reduce offline volume size without any special
flag and cause data loss, whereas online raising of volume size, which
should not be an issue, is refused by default, and does not seem possible
to force with `online=true`.

Originally-by: Guillaume Thouvenin <guillaume.thouvenin@vates.tech>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: activate default implementation for Volume.set_name and set_description

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: implement Volume.snapshot 1/n

Snapshot creation

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: implement Volume.snapshot 2/n

SR.ls for `xe sr-scan`

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
WIP zfs-ng: implement Volume.snapshot 3/n

Volume.stat e.g. for `xe vdi.destroy` or `xe vdi-copy into-vdi-uuid=`

FIXME: `xe vdi-copy` then assumes we have a blockdev available for the
snapshot, which is wrong - we'd have to either stream the data or create
a clone volume.

ydirson committed:
zfs-ng: implement Volume.clone 1/n

"Clone a snapshot" only, especially needed for `xe snapshot-revert`.

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: zfs-ng: implement Volume.clone 2/n

Allow destroying a VDI when it has been cloned, by first "zfs-promoting"
each of those clones to remove all dependencies on it.

Makes sure to log the pool state before and after promotions, to be able
to understand any problem with this logic.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: implement Volume.snapshot 4/n

Allow destruction of a snapshot with no dependency.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: fork xapi.storage.common.call() to shoehorse retry-on-busy

This is obviously not a long-term solution, we should transition to call
into the official zfs python lib instead of spawning processes.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: activate retry on busy on commands doing write operations

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
WIP zfs-ng: implement Volume.snapshot 5/n

When a snapshot with no child clones would block volume destruction,
create such a child clone.

FIXME: nothing will ever attempt to destroy such child clones, some sort
of GC is required.

gthvn1 committed:
Activate can be called several times

As it is idempotent we just need to return if the VDI is already
activate.

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

ydirson committed:
metabase: log VDI creations

This gives us a single log where all VDIs are registered, to get the
UUID / volume_id mapping for quicker debugging.

ydirson committed:
zfs-ng: implement sr-create 3/4: store pool name in meta.json

This will be required by `SR.stat`, which is needed by `xe sr-plug`.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: implement sr-create 4/4: SR.stat

This allows the "PDB plug on create" to work and completes the `xe
sr-create` implementation.

Note that since we cannot unplug a PBD yet the "xe sr-forget && zpool
destroy" we used up till now cannot work any more until that part is
done.

This is also another mandatory method needed by "xe sr-destroy".

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: implement sr-create 2/4: SR.attach

This is the first mandatory method needed by `xe sr-plug` and
`xe sr-destroy`.

`xe sr-plug` still does not complete, as it makes an `SR.stat` call
afterwards.

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: implement sr-unplug: SR.detach

This is also another mandatory method needed by `xe sr-destroy`.

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: implement sr-destroy: SR.destroy

Essentials of the SR lifecycle is now done.

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: implement sr-param-set: add SR.set_name and SR.set_description

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: implement volume-create 1/2: Volume.set and .unset

This activates the default implementations from libcow (insertion in
vdi_custom_keys table in SR metabase).  `Volume.set` is necessary for
`xe vdi-create`, and `Volume.unset` is provided for symmetry.

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: implement volume-create 2/2: Volume.create

Note that if Volume.set were not implemented yet (or indeed cannot
complete and returns an error), we face xapi-project/xen-api#5530: if
the Volume.set call fails (to add the vdi-type=user custom-key
record), then the VDI object is not created (though, as this happens
after Volume.create, the zvol and DB entries were indeed created and
there suppression was not requested by SMAPI after the error).  But
since all the VDI information that count are here, "xe sr-scan" will
bring those VDI to life (after implementation of SR.ls).

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: implement sr-scan: SR.ls

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: implement Volume.stat

This is necessary e.g. for `xe vdi-copy into-vdi-uuid=`

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-ng: expose device path for access to data

This is needed volume copy, `Datapath.attach` for `xe vbd-plug`, and such.

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

gthvn1 commented:
Just returning if `vdi.active_on` to be able to activate VDI twice is not enough because there is a refcounter on the `meta.pickle` file. It is the IntelliCache in the datapath. So at least we need to upgrade the refcounter but it is not just a counter so I'm trying to understand what we can and cannot do in `datapath.py:activate`

ydirson committed:
fixup! zfs-ng: implement sr-create 1/4: SR.create

ydirson committed:
fixup! zfs-ng: base infrastructure

ydirson committed:
fixup! zfs-ng: implement volume-create 2/2: Volume.create

ydirson commented on xapi/storage/libs/libcow/datapath.py:
```diff
@@ -53,11 +53,10 @@ def activate(cls, dbg, uri, domain, cb):
                     if vdi.active_on:
-                        raise util.create_storage_error(
-                            "SR_BACKEND_FAILURE_24",
-                            ["VDIInUse", "The VDI is currently in use"])
+                        return
```
Shouldn't we still raise an error if the VDI is active on a different domain than the requested one?

gthvn1 replied:
not sure if it is possible. It looks scary but before this I didn't think that we could plug two VBD for the same VDI... @Wescoeur what should we do? ideas?

ydirson commented on xapi/storage/libs/libcow/datapath.py:
```diff
@@ -53,11 +53,10 @@ def activate(cls, dbg, uri, domain, cb):
                 with cb.db_context(opq) as db:
                     vdi = db.get_vdi_by_id(key)
-                    # Raise Storage Error VDIInUse - 24
+                    # activate is idempotent. So just return if the VDI is
```
"*should be* idempotent"?

ydirson committed:
fixup! zfs-ng: base infrastructure

ydirson committed:
fixup! zfs-ng: implement sr-scan: SR.ls

ydirson committed:
fixup! zfs-ng: implement sr-create 1/4: SR.create

ydirson committed:
zfs-vol: new mode of SR creation with a ZFS vdev specification

This will allow management layers to present a suitable UI for the cases
they see fit without having to modify the interface.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>
